picoquic/logwriter.c

```diff
@@ -736,38 +742,65 @@ void binlog_picotls_ticket(FILE* f, picoquic_connection_id_t cnx_id,
     (void)fwrite(bytestream_data(msg), bytestream_length(msg), 1, f);
 }
 
+FILE* create_binlog(char const* binlog_file, uint64_t creation_time);
+
 void binlog_new_connection(picoquic_cnx_t * cnx)
 {
-    FILE * f = cnx->quic->f_binlog;
-    if (f == NULL) {
+    if (cnx == NULL || cnx->quic->binlog_dir == NULL) {
         return;
     }
 
-    bytestream_buf stream_msg;
-    bytestream * msg = bytestream_buf_init(&stream_msg, BYTESTREAM_MAX_BUFFER_SIZE);
-    bytewrite_cid(msg, &cnx->initial_cnxid);
-    bytewrite_vint(msg, cnx->start_time);
-    bytewrite_vint(msg, picoquic_log_event_new_connection);
+    int ret = 0;
 
-    bytewrite_int8(msg, cnx->client_mode != 0);
-    bytewrite_int32(msg, cnx->proposed_version);
-    bytewrite_cid(msg, &cnx->path[0]->remote_cnxid);
+    cnx->f_binlog = picoquic_file_close(cnx->f_binlog);
+    
+    char cid_name[2 * PICOQUIC_CONNECTION_ID_MAX_SIZE + 1];
+    if (picoquic_print_connection_id_hexa(cid_name, sizeof(cid_name), &cnx->initial_cnxid) != 0) {
+        ret = -1;
+    }
 
-    /* Algorithms used */
-    bytewrite_cstr(msg, cnx->congestion_alg->congestion_algorithm_id);
-    bytewrite_vint(msg, cnx->spin_policy);
+    char log_filename[512];
+    if (ret == 0) {
+        if (picoquic_sprintf(log_filename, sizeof(log_filename), NULL, "%s%s%s.log",
+            cnx->quic->binlog_dir, PICOQUIC_FILE_SEPARATOR, cid_name) != 0) {
+            ret = -1;
+        }
+    }
 
```

Comment:
This is good, but it also is the reason why the "packet_trace" test is failing. The test assumes that the binlog is in the file PACKET_TRACE_BIN in the current directory. With the new API, the trace would be in the folder PACKET_TRACE_BIN, with a name derived from the initial CID. Same problem for the "app_limit_cc" test: the code in `app_limit_cc_test_one` assumes that the trace is in APP_LIMIT_TRACE_BIN.

In both cases, the solution is to create the log in the current working directory -- set the folder name to "." or either "./" or ".\" depending on OS, derive a dynamic file name from the initial CID of `test_ctx->cnx_client`, and use that instead of the constant value.

There are more tests that generate bin logs today, and those are probably failing to generate these logs as well. I can deal with that later.

Reply:
Looking further, there is another alternative: add a parameter to the test context creation to use a predictable initial CID, much like what is done in the `qlog_trace` test. This is done in an ad-hoc way in the qlog test, delete the created connection and then create a new one, but that could be simplified by passing the initial CID as a parameter in (tls_api_init_ctx), so it could then be used in that code when creating the client connection context.